<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://CDPHE-bioinformatics.github.io/CDPHE-SARS-CoV-2/assembly_workflow/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Assembly Workflows - CDPHE-SARS-CoV-2</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Assembly Workflows";
        var mkdocs_page_input_path = "assembly_workflow.md";
        var mkdocs_page_url = "/CDPHE-SARS-CoV-2/assembly_workflow/";
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> CDPHE-SARS-CoV-2
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../setup/">Setup</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../python_scripts/">Python Scripts</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Assembly Workflows</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#illumina-pe">Illumina PE</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#1-terra-data-table">1. Terra Data Table</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#2-terra-workspace-data">2. Terra Workspace Data</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#3-setting-up-the-workflow-inputs">3. Setting Up the Workflow Inputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#illumina-se">Illumina SE</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_1">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#1-terra-data-table_1">1. Terra Data Table</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#2-terra-workspace-data_1">2. Terra Workspace Data</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_1">Outputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#1-output-files-from-trimmomatic-and-ddbuk">1. Output Files from Trimmomatic and ddbuk</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#2-output-files-from-fastqc">2. Output Files from FastQC</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#3-output-files-from-bwa-and-samtools-align-reads">3. Output files from bwa and samtools (align reads)</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#4-output-files-from-ivar-trim-and-samtools">4. Output files from iVar trim and samtools</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#5-output-files-from-ivar-variants">5. Output files from iVar variants</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#6-output-files-from-ivar-consensus">6. Output files from iVar consensus</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#7-output-files-from-samtools-flagstat-stats-and-percent_coverage">7. Output files from Samtools flagstat, stats, and percent_coverage</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#8-output-from-rename-consensus-fasta-headers">8. Output from rename consensus fasta headers</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#9-output-from-calc_percent_coveragepy">9. Output from calc_percent_coverage.py</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#10-bwa-assembler-version-string-output">10. bwa assembler version string output</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#oxford-nanopore-technologies-ont">Oxford Nanopore Technologies (ONT)</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_2">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#1-terra-data-table_2">1. Terra Data Table</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#2-terra-workspace-data_2">2. Terra Workspace Data</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#3-setting-up-the-workflow-inputs_1">3. Setting Up the Workflow Inputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_2">Outputs</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../transfer_workflow/">Transfer Workflows</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../results_workflow/">Results Workflows</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../wastewater_workflow/">Wastewater Workflow</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../nextstrain_workflow/">Nextstrain Workflow</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">CDPHE-SARS-CoV-2</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li>Assembly Workflows</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="assembly-workflows">Assembly Workflows</h1>
<p>The following three workflows describe the reference based assembly methods for paired-end and single end illumina seuqencing data and ONT sequencing data. Each workflow accepts "sample" as the root entity type.</p>
<h2 id="illumina-pe">Illumina PE</h2>
<p>File: SC2_illumina_pe_assembly.wdl</p>
<p>This workflow was developed for the assembly of Illumina 150 bp paired-end read data using the Illumina Nextera XT library prep protocol. The workflow accepts "sample" as the root entity type. The workflow will:</p>
<ol>
<li>Use Seqyclean to quality filter and trim raw fastq files<ul>
<li>Seqyclean parameters include a minimum read length set to 70 bp and quality trimming set to a minimum Phred quality score of 30.</li>
</ul>
</li>
<li>Run FastQC on both the raw and cleaned reads</li>
<li>Align reads to the reference genome using bwa and then sort the bam by coordinates using Samtools</li>
<li>Use iVar trim to trim primer regions and then sort the trimmed bam by coordinates using Samtools</li>
<li>Use iVar variants to call variants from the trimmed and sorted bam<ul>
<li>iVar variants parameters include a minimum quality score set to 20, a minimum variant base frequency set to 0.6 and a minimum read depth set to 10.</li>
</ul>
</li>
<li>Use iVar consensus to call the consensus genome sequence from the trimmed and sorted bam<ul>
<li>iVar consensus parameters include a minimum quality score set to 20, a minimum variant base frequency set to 0.6 and a minimum read depth set to 10.</li>
</ul>
</li>
<li>Use Samtools flagstat, stats, and coverage to output statistics from the bam</li>
<li>Rename the fasta header of consensus sequences in the GISAID-acceptable format: CO-CDPHE-{sample_id}</li>
<li>Calculate the percent coverage using the <code>calc_percent_coverage.py</code> script available in the <a href="https://github.com/CDPHE-bioinformatics/CDPHE-SARS-CoV-2/tree/main/workflows/python_scripts">python_scripts</a> directory of this repo.</li>
</ol>
<p><img alt="SC2_illumina_pe_assembly.wdl workflow diagram" src="../img/SC2_illumina_pe_assembly.png" /></p>
<h3 id="inputs">Inputs</h3>
<h4 id="1-terra-data-table">1. Terra Data Table</h4>
<p>The terra data table must include the following columns as listed below. Note that optional columns are not neccessary for the assembly workflow but must be present for the SC2_lineage_calling_and results.wdl and Transfer workflows described below under <code>Lineage Calling Workflows</code> and <code>Transfer Workflows</code>, respecitively.</p>
<table>
<thead>
<tr>
<th>column header</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>entity:sample_id</code></td>
<td>column with the list of sample names. (e.g. <code>entity:covwwt-0203_id</code>)</td>
</tr>
<tr>
<td><code>fastq_1</code></td>
<td>The google bucket path to the R1 fastq file.</td>
</tr>
<tr>
<td><code>fastq_2</code></td>
<td>The google bucket path to the R2 fastq file.</td>
</tr>
<tr>
<td><code>out_dir</code></td>
<td>User defined google bucket for where the files will be transfered during the transfer workflows.</td>
</tr>
<tr>
<td><code>workbook_path</code></td>
<td>(optional; required for lineage calling workflow)</td>
</tr>
<tr>
<td><code>project_name</code></td>
<td>(optional; requried for lineage calling workflow)</td>
</tr>
</tbody>
</table>
<h4 id="2-terra-workspace-data">2. Terra Workspace Data</h4>
<p>See <a href="../setup/">setup</a>.</p>
<h4 id="3-setting-up-the-workflow-inputs">3. Setting Up the Workflow Inputs</h4>
<p>For setting up the worklfow inputs, use the <code>SC2_illumina_pe_assembly-input.json</code> in the <code>workflow_inputs</code> directory.</p>
<table>
<thead>
<tr>
<th>workflow variable</th>
<th>attribute (input syntax into workflow)</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adapters_and_contaminants</code></td>
<td>workspace.adapters_and_contaminants_fa</td>
</tr>
<tr>
<td><code>calc_percent_coverage_py</code></td>
<td>workspace.covid_calc_percent_coverage_py</td>
</tr>
<tr>
<td><code>covid_genome</code></td>
<td>workspace.covid_genome_fa</td>
</tr>
<tr>
<td><code>covid_gff</code></td>
<td>workspace.covid_genome_gff</td>
</tr>
<tr>
<td><code>fastq_1</code></td>
<td>this.fastq_1</td>
</tr>
<tr>
<td><code>fastq_2</code></td>
<td>this.fastq_2</td>
</tr>
<tr>
<td><code>primer_bed</code></td>
<td>workspace.artic_v4-1_bed</td>
</tr>
<tr>
<td><code>s_gene_amplicons</code></td>
<td>workspace.artic_v4-1_s_gene_amplicons</td>
</tr>
<tr>
<td><code>sample_name</code></td>
<td>this.{entity_name}_id</td>
</tr>
</tbody>
</table>
<h3 id="outputs">Outputs</h3>
<table>
<thead>
<tr>
<th>WDL task name</th>
<th>software/program</th>
<th>variable name</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>seqyclean</td>
<td>seqyclean</td>
<td><code>filtered_reads_1</code></td>
<td>file</td>
</tr>
<tr>
<td>seqyclean</td>
<td>seqyclean</td>
<td><code>filtered_reads_2</code></td>
<td>file</td>
</tr>
<tr>
<td>seqyclean</td>
<td>seqyclean</td>
<td><code>seqyclean_summary</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_raw</td>
<td>fastqc</td>
<td><code>fastqc_raw1_html</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_raw</td>
<td>fastqc</td>
<td><code>fastqc_raw1_zip</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_raw</td>
<td>fastqc</td>
<td><code>fastqc_raw2_html</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_raw</td>
<td>fastqc</td>
<td><code>fastqc_raw2_zip</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_cleaned</td>
<td>fastqc</td>
<td><code>fastqc_clean1_html</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_cleaned</td>
<td>fastqc</td>
<td><code>fastqc_clean1_zip</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_cleaned</td>
<td>fastqc</td>
<td><code>fastqc_clean2_html</code></td>
<td>file</td>
</tr>
<tr>
<td>fastqc as fastqc_cleaned</td>
<td>fastqc</td>
<td><code>fastqc_clean2_zip</code></td>
<td>file</td>
</tr>
<tr>
<td>align_reads</td>
<td>bwa and samtools</td>
<td><code>out_bam</code></td>
<td>file</td>
</tr>
<tr>
<td>align_reads</td>
<td>bwa and samtools</td>
<td><code>out_bamindex</code></td>
<td>file</td>
</tr>
<tr>
<td>align_reads</td>
<td>bwa and samtools</td>
<td><code>assembler_version</code></td>
<td>string recording the version for bwa, this information is used later for submitting to public repositories.</td>
</tr>
<tr>
<td>ivar trim</td>
<td>ivar trim and samtools</td>
<td><code>trim_bam</code></td>
<td>file</td>
</tr>
<tr>
<td>ivar trim</td>
<td>ivar trim and samtools</td>
<td><code>trimsort_bam</code></td>
<td>file</td>
</tr>
<tr>
<td>ivar trim</td>
<td>ivar trim and samtools</td>
<td><code>trimsort_bamindex</code></td>
<td>file</td>
</tr>
<tr>
<td>ivar variants</td>
<td>ivar variants</td>
<td><code>variants</code></td>
<td>vcf file formatted as a tsv</td>
</tr>
<tr>
<td>ivar consensus</td>
<td>ivar consnesus</td>
<td><code>consensus</code></td>
<td>fasta file of consensus genome, Ns are called in places with less than 10 bp read depth.</td>
</tr>
<tr>
<td>bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>flagstat_out</code></td>
<td>file</td>
</tr>
<tr>
<td>bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>stats_out</code></td>
<td>file</td>
</tr>
<tr>
<td>bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>covhist_out</code></td>
<td>file</td>
</tr>
<tr>
<td>bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>cov_out</code></td>
<td>file</td>
</tr>
<tr>
<td>bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>cov_s_gene_amplcions_out</code></td>
<td>file</td>
</tr>
<tr>
<td>bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>cov_s_gene_out</code></td>
<td>file</td>
</tr>
<tr>
<td>rename_fasta</td>
<td>N/A</td>
<td><code>renamed_consensus</code></td>
<td>fasta file; consensus genome sequence with the fasta header renamed to be CO-CDPHE-{sample_name}</td>
</tr>
<tr>
<td>calc_percent_cvg</td>
<td>calc_percent_coverage.py</td>
<td><code>percent_cvg_csv</code></td>
<td>csv file, see calc_percent_cvg.py script readme for details found in the ./python_scripts directory of this repository.</td>
</tr>
</tbody>
</table>
<h2 id="illumina-se">Illumina SE</h2>
<p><strong>This workflow is no longer maintained.</strong></p>
<p>File: SC2_illumina_se_assembly.wdl</p>
<p>This workflow was developed for the assembly of Illumina 72 bp single-end read data using the Illumina COVIDSEQ library prep protocol. The workflow accepts "sample" as the root entity type. The workflow will:</p>
<ol>
<li>Use Trimmomatic and bbduk to quality filter, trim, and remove adapters from raw fastq files<ul>
<li>Trimmomatic parameters inlcude a sliding widnow set to trim reads when the 4bp sliding window quality score falls below a mean Phred quality score of 30 (i.e. 4:30) and a minimum read lenght of 25 bp.</li>
<li>bbduck parameters include adapter trimming set to trim everything to the right of a kmer match and removal of PhiX sequences.</li>
</ul>
</li>
<li>Run FastQC on both the raw and cleaned reads</li>
<li>Align reads to the reference genome using bwa and then sort the bam by coordinates using Samtools</li>
<li>Use iVar trim to trim primer regions and then sort the trimmed bam by coordinates using Samtools</li>
<li>Use iVar variants to call variants from the trimmed and sorted bam<ul>
<li>iVar variants parameters include a minimum quality score set to 20, a minimum variant base frequency set to 0.6 and a minimum read depth set to 10.</li>
</ul>
</li>
<li>Use iVar consensus to call the consensus genome sequence from the trimmed and sorted bam<ul>
<li>iVar consensus parameters include a minimum quality score set to 20, a minimum variant base frequency set to 0.6 and a minimum read depth set to 10.</li>
</ul>
</li>
<li>Use Samtools flagstat, stats, and coverage to output statistics from the bam</li>
<li>Rename the fasta header of consensus sequences in the format: CO-CDPHE{sample_id}</li>
<li>Calculate the percent coverage using the <code>calc_percent_coverage.py</code> script available in the <a href="https://github.com/CDPHE-bioinformatics/CDPHE-SARS-CoV-2/tree/main/workflows/python_scripts">python_scripts</a> directory of this repo.</li>
</ol>
<p><img alt="SC2_illumina_se_assembly.wdl workflow diagram" src="../img/SC2_illumina_se_assembly.png" /></p>
<h3 id="inputs_1">Inputs</h3>
<h4 id="1-terra-data-table_1">1. Terra Data Table</h4>
<p>The terra data table can be generated using the preprocess python scripts available in the <a href="https://github.com/CDPHE-bioinformatics/seq_data_preprocessing-private">data preprocessing repository</a>. The terra data table must include the following columns as listed below. Note that optional columns are not neccessary for the assembly workflow but be present for the SC2_lineage_calling_and results.wdl and Transfer workflows described below under <code>Lineage Calling Workflows</code> and <code>Transfer Workflows</code>, respectively.</p>
<ol>
<li><code>entity:sample_id</code>: column with the list of sample names/ids. Note that if there is more than one data table in the Terra Workspace, you need to add a number after the word sample to keep the datatables seperate (e.g. <code>entity:sample2_id</code>).</li>
<li><code>fastq</code>: The google bucket path to the fastq file.</li>
<li><code>seq_run</code> (optional): the name of the sequencing run (e.g. NEXSEQ_101)</li>
<li><code>tech_platform</code> (optional) : e.g. Illumina NexSeq</li>
<li><code>read_type</code> (optional): single</li>
<li><code>primer_set</code> (optional): e.g. COVIDSeqV3</li>
<li><code>plate_name</code> (optional): name of sequencing plate  </li>
<li><code>plate_sample_well</code> (optional): location of well on sequencing plate</li>
<li><code>out_dir</code> (optional): user defined google bucket fro where the files will be transfered during the transfer workflows.</li>
</ol>
<h4 id="2-terra-workspace-data_1">2. Terra Workspace Data</h4>
<p>The following reference files can be found in the <a href="https://github.com/CDPHE-bioinformatics/CDPHE-SARS-CoV-2/tree/main/workflows/workspace_data">workspace_data</a> directory and the <a href="https://github.com/CDPHE-bioinformatics/CDPHE-SARS-CoV-2/tree/main/workflows/python_scripts">python_scripts</a> directory. These files should be saved as Workspace data in your Terra Workspace. To do so, upload the files to a google bucket an link the file path to the wrokspace data variable. Once saved as workspace data variables, they can be used as inputs for the workflow.</p>
<ol>
<li><code>covid_genome</code>: the path to the google bucket directory contianing the SARS-CoV-2 reference genome fasta (we use NCBI genbank ID MN908947.3).</li>
<li><code>covid_gff</code>: the path to the google bucket directory containing the SARS-CoV-2 reference genome gff annotation file (we use NCBI genbank ID MN908947.3)</li>
<li><code>primer_bed</code>: the path to the google bucket directory containing a bed file with the primers used for amplicon sequencing<ul>
<li>currenly we have bed files for Artic V3, Artic V4, Artic V4.1 and Midnight.</li>
</ul>
</li>
<li><code>preprocess_python_script</code>: [do we want to change the name of this variable in the WDL to match the python script name?] the path to the google bucket containing the <code>calc_percent_coverage.py</code> script.</li>
</ol>
<p>Below is a summary of the workflow input variables along with the syntax used for the attribute column when setting up the workflow to run on Terra.bio. For the attributes, the "this." syntax refers Terra to pull the variable from the terra datatable (#1 above). The "workspace." syntax refers Terra to pull the variable from the terra workspace data (#2 above).</p>
<table>
<thead>
<tr>
<th>workflow variable</th>
<th>attribute (input syntax into workflow)</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>covid_genome</code></td>
<td>workspace.covid_genome</td>
</tr>
<tr>
<td><code>covid_gff</code></td>
<td>workspace.covid_gff</td>
</tr>
<tr>
<td><code>fastq</code></td>
<td>this.fastq</td>
</tr>
<tr>
<td><code>preprocess_python_script</code></td>
<td>workspace.preprocess_python_script</td>
</tr>
<tr>
<td><code>primer_bed</code></td>
<td>workspace.V4-1Artic</td>
</tr>
<tr>
<td><code>sample_id</code></td>
<td>this.sample{terra_datatable_name}_id</td>
</tr>
</tbody>
</table>
<h3 id="outputs_1">Outputs</h3>
<h4 id="1-output-files-from-trimmomatic-and-ddbuk">1. Output Files from Trimmomatic and ddbuk</h4>
<ul>
<li><code>trimmed_reads</code>: file</li>
<li><code>trim_stats</code>: file</li>
<li><code>filtered_reads</code>: file</li>
<li><code>adapter_stats</code>: file</li>
<li><code>PhiX_stats</code>: file</li>
</ul>
<h4 id="2-output-files-from-fastqc">2. Output Files from FastQC</h4>
<ul>
<li><code>fastqc_raw1_html</code>: file</li>
<li><code>fastqc_raw1_zip</code>: file</li>
<li><code>fastqc_raw2_html</code>: file</li>
<li><code>fastqc_raw2_zip</code>: file</li>
<li><code>fastqc_clean1_html</code>: file</li>
<li><code>fastqc_clean1_zip</code>: file</li>
<li><code>fastqc_clean2_html</code>: file</li>
<li><code>fastqc_clean2_zip</code>: file</li>
</ul>
<h4 id="3-output-files-from-bwa-and-samtools-align-reads">3. Output files from bwa and samtools (align reads)</h4>
<ul>
<li><code>out_bam</code>: file</li>
</ul>
<h4 id="4-output-files-from-ivar-trim-and-samtools">4. Output files from iVar trim and samtools</h4>
<ul>
<li><code>trim_bam</code>: file</li>
<li><code>trimsort_bam</code>: file</li>
<li><code>trimsort_bamindex</code>: file</li>
</ul>
<h4 id="5-output-files-from-ivar-variants">5. Output files from iVar variants</h4>
<ul>
<li><code>variants</code>: vcf file formated as a tsv</li>
</ul>
<h4 id="6-output-files-from-ivar-consensus">6. Output files from iVar consensus</h4>
<ul>
<li><code>consensus</code>: fasta file of conensus genome, Ns are called in places with less than 10 bp read depth. </li>
</ul>
<h4 id="7-output-files-from-samtools-flagstat-stats-and-percent_coverage">7. Output files from Samtools flagstat, stats, and percent_coverage</h4>
<ul>
<li><code>fagstat_out</code>: file</li>
<li><code>stats_out</code>: file</li>
<li><code>covhist_out</code>: file</li>
<li><code>cov_out</code>: file</li>
</ul>
<h4 id="8-output-from-rename-consensus-fasta-headers">8. Output from rename consensus fasta headers</h4>
<ul>
<li><code>renamed_consensus</code>: fasta file; consesnus genome sequence with the fasta header renamed to be CO-CDPHE-{sample_id}</li>
</ul>
<h4 id="9-output-from-calc_percent_coveragepy">9. Output from calc_percent_coverage.py</h4>
<ul>
<li><code>percent_cvg_csv</code>: csv file, see calc_percent_cvg.py script readme for details.</li>
</ul>
<h4 id="10-bwa-assembler-version-string-output">10. bwa assembler version string output</h4>
<ul>
<li><code>assembler_version</code>: string recording the version for bwa, this information is used later for submitting to public repositories.</li>
</ul>
<h2 id="oxford-nanopore-technologies-ont">Oxford Nanopore Technologies (ONT)</h2>
<p>File: SC2_ont_assembly.wdl</p>
<p>This workflow was developed for the assembly of Oxford Nanopore Technology (ONT) read data following the ARTIC SARS-CoV-2 sequencing protocol and using the ONT native barcoding kit. This workflow assumes that basecalling and conversion of fast5 files into fastq has already occurred (e.g. using MinKNOW). The workflow accepts "sample" as the root entity type. The workflow will:</p>
<ol>
<li>Demuliplex basecalled fastq files using guppy_barcoder</li>
<li>Perform quality filering using guppyplex<ul>
<li>guppyplex inlcudes a min length parameter set to 400 and a max length set to 700 for Artic primers and a min lingth set ot 400 and a max length set to 1500 for midnight primers.</li>
</ul>
</li>
<li>Run artic minion --medaka for variant calling and to generate a consensus fagstat_out<ul>
<li>medaka uses minimap2 by default to align reads to the SARS-CoV-2 reference genome</li>
<li>the default parameter in medaka for base calling is 20x depth and at least 60% of reads containing the base call</li>
</ul>
</li>
<li>Scaffold assembly with pyScaf<ul>
<li>this step ensures a single continuous consensus sequence with only one sequence in the consensus fasta file</li>
</ul>
</li>
<li>Rename consensus to CO-CDPHE-{sample_id}</li>
<li>Generate bam quality statistics using samtools</li>
<li>Calculates percent coverage using the <code>calc_percent_coverage.py</code> script</li>
</ol>
<p><img alt="SC2_ont_assembly.wdl workflow diagram" src="../img/SC2_ont_assembly.png" /></p>
<h3 id="inputs_2">Inputs</h3>
<h4 id="1-terra-data-table_2">1. Terra Data Table</h4>
<p>The terra data table can be generated using the preprocess python scripts available in the <a href="https://github.com/CDPHE-bioinformatics/seq_data_preprocessing-private">data preprocessing repository</a>. The terra data table must include the following columns as listed below. Note that optional columns are not neccessary for the assembly workflow but but be present for the SC2_lineage_calling_and results.wdl and Transfer workflows described below under <code>Lineage Calling Workflows</code> and <code>Transfer Workflows</code>, respectively.</p>
<table>
<thead>
<tr>
<th>column header</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>entity:sample_id</code></td>
<td>column with the list of sample names. (e.g. <code>entity:covwwt-0203_id</code>)</td>
</tr>
<tr>
<td><code>index_1_id</code></td>
<td>the ont barcode associated with the sample</td>
</tr>
<tr>
<td><code>fastq_dir</code></td>
<td>the google bucket path with the set of fastq files</td>
</tr>
<tr>
<td><code>out_dir</code></td>
<td>User defined google bucket for where the files will be transfered during the transfer workflows.</td>
</tr>
<tr>
<td><code>workbook_path</code></td>
<td>(optional; required for lineage calling workflow)</td>
</tr>
<tr>
<td><code>project_name</code></td>
<td>(optional; requried for lineage calling workflow)</td>
</tr>
</tbody>
</table>
<h4 id="2-terra-workspace-data_2">2. Terra Workspace Data</h4>
<p>See <a href="../setup/">setup</a>.</p>
<h4 id="3-setting-up-the-workflow-inputs_1">3. Setting Up the Workflow Inputs</h4>
<p>For setting up the worklfow inputs, use the <code>SC2_ont_assembly-input.json</code> in the <code>workflow_inputs</code> directory.</p>
<table>
<thead>
<tr>
<th>workflow variable</th>
<th>attribute (input syntax into workflow)</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>calc_percent_coverage_py</code></td>
<td>workspace.covid_calc_percent_coverage_py</td>
</tr>
<tr>
<td><code>covid_genome</code></td>
<td>workspace.covid_genome_fa</td>
</tr>
<tr>
<td><code>gcs_fastq_dir</code></td>
<td>this.fastq_dir</td>
</tr>
<tr>
<td><code>index_1_id</code></td>
<td>this.index_1_id</td>
</tr>
<tr>
<td><code>primer_bed</code></td>
<td>workspace.artic_v4-1_bed</td>
</tr>
<tr>
<td><code>primer_set</code></td>
<td>this.primer_set</td>
</tr>
<tr>
<td><code>s_gene_amplicons</code></td>
<td>workspace.artic_v4-1_s_gene_amplicons</td>
</tr>
<tr>
<td><code>s_gene_primer_bed</code></td>
<td>workspace.artic_v4-1_s_gene_primer_bed</td>
</tr>
<tr>
<td><code>sample_name</code></td>
<td>this.{entity_name}_id</td>
</tr>
</tbody>
</table>
<h3 id="outputs_2">Outputs</h3>
<table>
<thead>
<tr>
<th>WDL task name</th>
<th>software/program</th>
<th>variable name</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Demultiplex</td>
<td>guppy_barcoder</td>
<td><code>barcode_summary</code></td>
<td>file</td>
</tr>
<tr>
<td>Demultiplex</td>
<td>guppy_barcoder</td>
<td><code>guppy_dmux_fastq</code></td>
<td>file</td>
</tr>
<tr>
<td>guppyplex quality filtering</td>
<td><code>filtered_fatsq</code></td>
<td>file</td>
<td></td>
</tr>
<tr>
<td>Medaka</td>
<td>medaka and minimap2</td>
<td><code>sorted_bam</code></td>
<td>file</td>
</tr>
<tr>
<td>Medaka</td>
<td>medaka and minimap2</td>
<td><code>trim_sort_bam</code></td>
<td>file</td>
</tr>
<tr>
<td>Medaka</td>
<td>medaka and minimap2</td>
<td><code>trimsort_bai</code></td>
<td>file</td>
</tr>
<tr>
<td>Medaka</td>
<td>medaka and minimap2</td>
<td><code>variants</code></td>
<td>file</td>
</tr>
<tr>
<td>Medaka</td>
<td>medaka and minimap2</td>
<td><code>consensus</code></td>
<td>file</td>
</tr>
<tr>
<td>Medaka</td>
<td>medaka and minimap2</td>
<td><code>assembler_version</code></td>
<td>string recording the version for artic medaka, this information is used later for submitting to public repositories.</td>
</tr>
<tr>
<td>Bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>flagstat_out</code></td>
<td>file</td>
</tr>
<tr>
<td>Bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>stats_out</code></td>
<td>file</td>
</tr>
<tr>
<td>Bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>covhist_out</code></td>
<td>file</td>
</tr>
<tr>
<td>Bam_stats</td>
<td>samtools flagstat, stats, percent_coverage</td>
<td><code>cov_out</code></td>
<td>file</td>
</tr>
<tr>
<td>Scaffold</td>
<td>pyScaf</td>
<td><code>scaffold_consensus</code></td>
<td>consesnus sequence as a fasta file</td>
</tr>
<tr>
<td>rename_fasta</td>
<td>N/A</td>
<td><code>renamed_consensus</code></td>
<td>fasta file; consesnus genome sequence with the fasta header renamed to be CO-CDPHE-{sample_name}</td>
</tr>
<tr>
<td>calc_percent_cvg</td>
<td>calc_percent_coverage.py</td>
<td><code>percent_cvg_csv</code></td>
<td>csv file, see calc_percent_cvg.py script readme for details found in the ./python_scripts directory of this repository.</td>
</tr>
<tr>
<td>get_primer_site_variants</td>
<td>bcftools</td>
<td><code>primer_site_variants</code></td>
<td>file</td>
</tr>
</tbody>
</table>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../python_scripts/" class="btn btn-neutral float-left" title="Python Scripts"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../transfer_workflow/" class="btn btn-neutral float-right" title="Transfer Workflows">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../python_scripts/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../transfer_workflow/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
