<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://CDPHE-bioinformatics.github.io/CDPHE-SARS-CoV-2/results_workflow/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Results Workflows - CDPHE-SARS-CoV-2</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Results Workflows";
        var mkdocs_page_input_path = "results_workflow.md";
        var mkdocs_page_url = "/CDPHE-SARS-CoV-2/results_workflow/";
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> CDPHE-SARS-CoV-2
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../setup/">Setup</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../python_scripts/">Python Scripts</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../assembly_workflow/">Assembly Workflows</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../transfer_workflow/">Transfer Workflows</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Results Workflows</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#lineage-calling-and-results-workflow">Lineage Calling and Results Workflow</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#multifasta-lineage-calling-and-results-workflow">Multifasta Lineage Calling and Results Workflow</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_1">Inputs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_1">Outputs</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../wastewater_workflow/">Wastewater Workflow</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../nextstrain_workflow/">Nextstrain Workflow</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">CDPHE-SARS-CoV-2</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li>Results Workflows</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="results-workflows">Results Workflows</h1>
<p>The following workflows will perform clade and lineage assignment using Nextclade and Pangolin, respectively. The SC2_lineage_calling_and_results.wdl should be run following assembly with one of the three above reference-based assembly workflows. The SC2_mulitfasta_lineage_calling can be run on any multifasta file and is not dependent on any of the above workflows.  </p>
<h2 id="lineage-calling-and-results-workflow">Lineage Calling and Results Workflow</h2>
<p>File: SC2_lineage_calling_and_results.wdl</p>
<p>This workflow should be run following assembly with one of the three reference based assembly workflows. The workflow accepts "sample_set" as the root entity type and uses the data table from any of the three assembly workflows. All three assembly workflows (illumina pe, illumina se, and ont) are compatible with this workflow. Breifly the workflow performs the following:</p>
<ol>
<li>concatenates consensus sequences from the sequencing run into a single fasta file</li>
<li>runs panoglin on the concatenated fasta file</li>
<li>runs nextclade on the concatenated fasta file</li>
<li>parses the nextclade json file output using the <code>nextclade_json_parser.py</code> script which pulls out clade and nucleotide and animo acid changes information and converts it to a tabular format.</li>
<li>Concaenates sequencing assembly metrics (e.g. percent coverage, assembler version), lineage and clade information, and sequence metadata (e.g. plate name, sample well location) into a single csv file.</li>
<li>Generates a csv file with sequencing assembly metrics and lineage information that can be used to parse sequencing data into our LIMS.</li>
<li>Transfers intermediate and summary files to a user defined google bucket.  </li>
</ol>
<p><img alt="SC2_lineage_calling_and_results.wdl workflow diagram" src="../img/SC2_lineage_calling_and_results_workflow.png" /></p>
<h3 id="inputs">Inputs</h3>
<p>Below is a summary of the workflow input variables along with the syntax used for the attribute column when setting up the workflow to run on Terra.bio. For the attributes, the "this.sample{terra_datatable_name}s." syntax refers Terra to pull the variable from the terra datatable as used for sample sets. These variables were either in the original terra datatable as inputs for the assembly workflow (see referece based assembly workflow inputs sections above for more details) or added as outputs during the assemlby workflow (see reference based assembly workflow outputs sections for more details). The "workspace." syntax refers Terra to pull the variable from the terra workspace data. Workspace data is describe in the <code>Getting Started</code> drop down menu above.</p>
<table>
<thead>
<tr>
<th>workflow variable</th>
<th>attribute (input syntax into workflow)</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td><code>assembler_version_array</code></td>
<td>this.sample{terra_datatable_name}s.assembler_version</td>
<td></td>
</tr>
<tr>
<td><code>concat_seq_results_py</code></td>
<td>workspace.covid_concat_results_py</td>
<td></td>
</tr>
<tr>
<td><code>cov_out</code></td>
<td>this.sample{terra_datatable_name}s.cov_out</td>
<td></td>
</tr>
<tr>
<td><code>nextclade_json_parser_py</code></td>
<td>workspace.covid_nextclade_json_parser_py</td>
<td></td>
</tr>
<tr>
<td><code>out_dir_array</code></td>
<td>this.sample{terra_datatable_name}s.out_dir</td>
<td></td>
</tr>
<tr>
<td><code>percent_cvg_csv</code></td>
<td>this.sample{terra_datatable_name}s.percent_cvg_csv</td>
<td></td>
</tr>
<tr>
<td><code>project_name_array</code></td>
<td>this.sample{terra_datatable_name}s.project_name</td>
<td></td>
</tr>
<tr>
<td><code>renamed_consensus</code></td>
<td>this.sample{terra_datatable_name}s.renamed_consesnus</td>
<td></td>
</tr>
<tr>
<td><code>sample_name</code></td>
<td>this.sample{terra_datatable_name}s.sample{terra_datatable_name}_id</td>
<td></td>
</tr>
<tr>
<td><code>workbook_path_array</code></td>
<td>this.sample{terra_datatable_name}s.workbook_path</td>
<td></td>
</tr>
</tbody>
</table>
<h3 id="outputs">Outputs</h3>
<p>This workflow generates several output files which are transfered to the user defined user google bucket as defined by this.sample{terra_datatable_name}s.out_dir. The table below details each output. For more detailed regarding the values in each column for the outputs see either the software readmes or the readme for the specific python script as listed in the description.</p>
<table>
<thead>
<tr>
<th>output variable name</th>
<th>file_name</th>
<th>description</th>
<th>google bucket path</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>cat_fastas</code></td>
<td><code>concatenate_assemblies.fasta</code></td>
<td>all consesnus sequences from assembly in a single fasta file</td>
<td><code>gs://{user_defined_gcp_bucket}/multifasta/</code></td>
</tr>
<tr>
<td><code>nextclade_clades_csv</code></td>
<td><code>{seq_run}_nextclade_results.csv</code></td>
<td>csv file generated from the <code>nextclade_json_parser.py</code> script detailing the clade for each seqeunce</td>
<td><code>gs://{user_defined_gcp_bucket}/nextclade_out/</code></td>
</tr>
<tr>
<td><code>nextclade_csv</code></td>
<td><code>nextclade.csv</code></td>
<td>csv file generated from nextclade</td>
<td><code>gs://{user_defined_gcp_bucket}/nextclade_out/</code></td>
</tr>
<tr>
<td><code>nextclade_json</code></td>
<td><code>nextclade.json</code></td>
<td>json file generated from nextclade; this json file is parsed using the <code>nextclade_json_parser.py</code> script and key info is pulled out and converted into a tablular format in the <code>nextclade_clades_csv</code>, <code>nextclade_variants_csv</code> and <code>sequencing_results.csv</code> files (see the readme for the <code>nextclade_json_parser.py</code> script for more details)</td>
<td><code>gs://{user_defined_gcp_bucket}/nextclade_out/</code></td>
</tr>
<tr>
<td><code>nextclade_variants_csv</code></td>
<td><code>{seq_run}_nextclade_variant_summary.csv</code></td>
<td>csv file generated from the <code>nextclade_json_parser.py</code> script detailing the nucleotide and amino acid changes for each seqeunce</td>
<td><code>gs://{user_defined_gcp_bucket}/summary_results/</code></td>
</tr>
<tr>
<td><code>nextclade_version</code></td>
<td>N/A</td>
<td>version of nextclade</td>
<td>N/A</td>
</tr>
<tr>
<td><code>pangolin_lineage</code></td>
<td><code>pangolin_lineage_report.csv</code></td>
<td>lineage report generated from pangolin</td>
<td><code>gs://{user_defined_gcp_bucket}/pangolin/</code></td>
</tr>
<tr>
<td><code>pangolin_version</code></td>
<td>N/A</td>
<td>version of panoglin</td>
<td>N/A</td>
</tr>
<tr>
<td><code>sequencing_results_csv</code></td>
<td><code>{seq_run}_sequencing_results.csv</code></td>
<td>summary of the sequencing metrics and lineage/clade assignments for each sequence generated from the <code>concat_seq_metrics_and_lineage_results.py</code> script. see the <code>concat_seq_metrics_and_lineage_results.py</code> readme for more details.</td>
<td><code>gs://{user_defined_gcp_bucket}/summary_results/</code></td>
</tr>
<tr>
<td><code>wgs_horizon_report_csv</code></td>
<td><code>{seq_run}_wgs_horizon_report.csv</code></td>
<td>results csv used for parsing results into our LIMS. This file is generated from the <code>concat_seq_metrics_and_lineage_results.py</code> script. see the <code>concat_seq_metrics_and_lineage_results.py</code> readme for more details.</td>
<td><code>gs://{user_defined_gcp_bucket}/summary_results/</code></td>
</tr>
</tbody>
</table>
<h2 id="multifasta-lineage-calling-and-results-workflow">Multifasta Lineage Calling and Results Workflow</h2>
<p>File: SC2_multifasta_lineage_calling.wdl</p>
<p>This workflow will perfrom lineage and clade assignment using a concatenated fasta file as input. This workflow is a stand alone workflow and does not depend on any of the previous reference-based assembly workflows. The workflow accepts "sample" as the root entity type and uses a simple two column data table (see inputs below).</p>
<p>Brienfly this workflow will perform the following:</p>
<ol>
<li>Run nextclade</li>
<li>Run panoglin</li>
<li>transfer outputs to a user defined google bucket.</li>
</ol>
<p><img alt="SC2_multifasta_lineage_calling.wdl workflow diagram" src="../img/SC2_multifasta_lineage_calling.png" /></p>
<h3 id="inputs_1">Inputs</h3>
<ol>
<li>Terra datatable. You will need to create a terra data that contains the following two columns:<ol>
<li><code>entity:samle_id</code> : should contain the prefix that you would like for you output files</li>
<li><code>multifasta</code>: google bucket path to the multi-sequence fasta file</li>
</ol>
</li>
</ol>
<p>Below is a summary of the workflow input variables along with the syntax used for the attribute column when setting up the workflow to run on Terra.bio. For the attributes, the "this." syntax refers Terra to pull the variable from the terra datatable (#1 above).</p>
<table>
<thead>
<tr>
<th>workflow variable</th>
<th>attribute (input syntax into workflow)</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>multifasta</code></td>
<td>this.multifasta</td>
</tr>
<tr>
<td><code>sample_id</code></td>
<td>this.sample{terra_datatable_name}_id</td>
</tr>
<tr>
<td><code>out_dir</code></td>
<td>"gs://{path to user defined google bucket}"</td>
</tr>
</tbody>
</table>
<h3 id="outputs_1">Outputs</h3>
<p>This workflow generates several output files which are transfered to the user defined user google bucket as defined by the <code>out_dir</code> variable. The table below details each output.</p>
<table>
<thead>
<tr>
<th>output variable name</th>
<th>file_name</th>
<th>description</th>
<th>google bucket path</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>nextclade_version</code></td>
<td>N/A</td>
<td>nextclade version</td>
<td>N/A</td>
</tr>
<tr>
<td><code>nextclade_json</code></td>
<td><code>{sample_id}_nextclade_json</code></td>
<td>json file generated from nextclade</td>
<td><code>gs://{user_defined_gcp_bucket}/nextclade_out</code></td>
</tr>
<tr>
<td><code>auspice_json</code></td>
<td><code>{sample_id}_nextclade.auspice.json</code></td>
<td>auspice json file generated from nextclade</td>
<td><code>gs://{user_defined_gcp_bucket}/nextclade_out</code></td>
</tr>
<tr>
<td><code>nextclade_csv</code></td>
<td><code>{sample_id}_nextclade.csv</code></td>
<td>csv file generated from nextclade</td>
<td><code>gs://{user_defined_gcp_bucket}/nextclade_out</code></td>
</tr>
<tr>
<td><code>pangolin_version</code></td>
<td>N/A</td>
<td>pangolin version</td>
<td>N/A</td>
</tr>
<tr>
<td><code>pangolin_lineage</code></td>
<td><code>{sample_id}_panoglin_lineage_report.csv</code></td>
<td>lineage report generated from pangolin</td>
<td><code>gs://{user_defined_gcp_bucket}/pangolin_out</code></td>
</tr>
</tbody>
</table>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../transfer_workflow/" class="btn btn-neutral float-left" title="Transfer Workflows"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../wastewater_workflow/" class="btn btn-neutral float-right" title="Wastewater Workflow">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../transfer_workflow/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../wastewater_workflow/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
